src/AssetsContractController.ts

```diff
@@ -0,0 +1,193 @@
+
+	/**
+	 *
+	 * Query if a contract implements an interface
+	 *
+	 * @param address - Asset contract address
+	 * @param interfaceId - Interface identifier
+	 * @returns - Promise resolving to whether the contract implements `interfaceID`
+	 */
+	private async contractSupportsInterface(address: string, interfaceId: string): Promise<boolean> {
+		/* istanbul ignore if */
+		if (!this.web3) {
+			return false;
+		}
+		try {
+			const contract = this.web3.eth.contract(abiERC721).at(address);
+			return await new Promise<boolean>((resolve, reject) => {
+				contract.supportsInterface(interfaceId, (error: Error, result: boolean) => {
+					/* istanbul ignore if */
+					if (error) {
+						reject(error);
+						return;
+					}
+					resolve(result);
+				});
+			});
+		} catch (error) {
+			/* istanbul ignore next */
+			/* waiting for https://github.com/ethereum/web3.js/issues/1119 */
+			return false;
+		}
+	}
+
+	/**
+	 * Name of this controller used during composition
+	 */
+	name = 'AssetsContractController';
+
+	/**
+	 * Creates a AssetsContractController instance
+	 *
+	 * @param config - Initial options used to configure this controller
+	 * @param state - Initial state to set on this controller
+	 */
+	constructor(config?: Partial<AssetsContractConfig>, state?: Partial<BaseState>) {
+		super(config, state);
+		this.defaultConfig = {
+			provider: undefined
+		};
+		this.initialize();
+	}
+
+	/**
+	 * Sets a new provider
+	 *
+	 * @property provider - Provider used to create a new underlying Web3 instance
+	 */
+	set provider(provider: any) {
+		this.web3 = new Web3(provider);
+	}
+
+	/**
+	 * Query if contract implements ERC721Metadata interface
+	 *
+	 * @param address - ERC721 asset contract address
+	 * @returns - Promise resolving to whether the contract implements ERC721Metadata interface
+	 */
+	async contractSupportsMetadataInterface(address: string): Promise<boolean> {
+		return this.contractSupportsInterface(address, ERC721METADATA_INTERFACE_ID);
+	}
+
+	/**
+	 * Query if contract implements ERC721Enumerable interface
+	 *
+	 * @param address - ERC721 asset contract address
+	 * @returns - Promise resolving to whether the contract implements ERC721Enumerable interface
+	 */
+	async contractSupportsEnumerableInterface(address: string): Promise<boolean> {
+		return this.contractSupportsInterface(address, ERC721ENUMERABLE_INTERFACE_ID);
+	}
+
+	/**
+	 * Get balance or count for current account on specific asset contract
+	 *
+	 * @param address - Asset contract address
+	 * @param selectedAddress - Current account public address
+	 * @returns - Promise resolving to balance for current account on specific asset contract
+	 */
+	async getBalanceOf(address: string, selectedAddress: string): Promise<typeof BN> {
+		/* istanbul ignore if */
+		if (!this.web3) {
+			return new BN(0);
+		}
+		try {
+			const contract = this.web3.eth.contract(abiERC20).at(address);
+			return await new Promise<typeof BN>((resolve, reject) => {
```

Comment:
@bitpshr  yeah this is weird, I did this as a workaround for https://github.com/ethereum/web3.js/issues/1119 since the workaround mentioned there wasn't working. The issue is still open, let me know what you think.

Reply:
Ah, ok, I see. Good catch 👍